### DISCUSSION TITLE
Alpine Expression Error appears when using wire:navigate and pressing the back button in the browser

### DISCUSSION THREAD

**Discussion by wiyatnanto** [livewire/livewire#6374] [1. Help]
when switching pages using wire:navigate everything works fine, but when visiting some pages and trying to return to the previous page it makes alpine not work properly. thanks by the way

**Comment by joshhanley** [comment]
@wiyatnanto thanks for reporting! Any chance you could demonstrate the issue with a screencapture or demo repo with steps? Thanks!

**Reply by achmedislamic** [reply]
I successfully reproduce the error. Here is the repo:

https://github.com/achmedislamic/lw3-back-error

To reproduce:
1. On the User Table menu, there is an Alpine Component just for example.
2. To trigger the error, click detail button on any user.
3. Click back button from the browser.
4. Check browser console to see the error.

The error is like this.
![image](https://github.com/livewire/livewire/assets/25688207/ed4418d2-2017-4326-8829-a53f69359517)

**Reply by achmedislamic** [reply]
https://github.com/livewire/livewire/assets/25688207/b04628dd-8a80-4dcc-a57b-c59e188c4786

here is the video demonstrating how to reproduce the error.

The error only occur with wire:navigate.

**Comment by joshhanley** [comment]
@wiyatnanto thanks for reporting! Any chance you could demonstrate the issue with a screencapture or demo repo with steps? Thanks!

**Comment by robsontenorio** [comment]
@joshhanley  Have same issue.

1. Visit this page https://www.mary-ui.com/docs/components/datepicker
2. Then, visit  any page on main menu (all have wire:navigate)
3. Now, click browser back button
4. See a glitch on datepicker fields

**Datepicker** component code is here  https://github.com/robsontenorio/mary/blob/5811b25b0211987e25f1c23a69af7e32737b6cf5/src/View/Components/DatePicker.php#L54


On `x-init ` I am calling `flatpickr` javascript function to initialize fields.


Maybe related to upcoming docs from @calebporzio ? #6523

**Reply by joydeep-bhowmik** [reply]
how did u fixed it?

**Reply by robsontenorio** [reply]
It is not fixed. Even on latest 3.0.5

**Reply by cbaconnier** [reply]
I believe the issue is caused by a re-run of `x-data` when navigating with the back button. I think it should not re-run since the DOM is preserved.

Using Livewire 3.2.2 I have been able to find a fix for  the flatpickr issue by destroying the instance when navigating: `x-on:livewire:navigating.window="instance.destroy();"`

Keep in mind 
1) I just tried and it worked, I didn't do extended tests to see if there's other failing scenario
2) I do think of a failing scenario: when the flatpickr instance is outside the page content (eg: using `@persist`), it should not listen for this event otherwise, it will be destroyed too.
3) I didn't test with returning back multiple times. I don't know if `wire:navigate` preserve the DOM for all the previous pages.
4) I didn't test the forth button.

At the moment I will not use `wire:navigate` as I think it's not fully reliable, yet. That's why I didn't test this solution further. 

```blade
<div wire:ignore>
    <input
        x-data="{ value: @entangle($attributes->wire('model')), instance: undefined }"
        x-init="() => {
            $watch('value', value => instance.setDate(value, false));
            instance = flatpickr($refs.input);
        }"
        x-on:livewire:navigating.window="instance.destroy();"
        x-ref="input"
        type="text"
        x-bind:value="value"
       {{ $attributes }}
    />
</div>
```

**Reply by devront-it** [reply]
Does anyone know if there is a way to NOT preserve the DOM when navigating back? Life would be so much easier if navigating back behaved like "normal" wire:navigate. This could potentially fix all the issues.

**Comment by joydeep-bhowmik** [comment]
this  issue is still not fixed in  v 3.0.5

**Comment by cwkhawand** [comment]
Having the same error on Alpine.JS v3.13.2

**Comment by devront-it** [comment]
Same issue with L. v3.1.0. Tried everything I could imagine with code, no luck. Any updates on this? 🫶🏼

**Reply by joydeep-bhowmik** [reply]
Not fixed yet

**Comment by tabatii** [comment]
Same issue with livewire v3.2.2

**Comment by robsontenorio** [comment]
For future reference, the case of **flatpickr** with **"altInput".**

- As flatpickr docs says when using "altInput" it "hides your original input and **creates a new one.**"
- So when using **wire:navigate** and navigate **back on browser** it duplicates the input

To solve this I have used the trick from @cbaconnier , thanks!

```html
<div
    x-data="{ instance: flatpickr($refs.myInput, { altInput: true, altFormat: 'F j, Y'}) }"
    x-on:livewire:navigating.window="instance.destroy();"
>
   <input x-ref="myInput" class="input input-primary" />
</div>
```

**Reply by ZermattChris** [reply]
Remove all *wire:navigate* from your Blade files...
and also remove all *navigate:true* from redirects:
```
$this->redirect('/posts', navigate: true);
```
from my Livewire controllers. Flatpickr is now happy again.

**Comment by aftabkhaliq86** [comment]
anyone has fixed that issue ?

**Reply by joydeep-bhowmik** [reply]
What version of livewire are you using?

**Reply by aftabkhaliq86** [reply]
am using 3.2.1

**Reply by aftabkhaliq86** [reply]
![image](https://github.com/livewire/livewire/assets/129963183/cb7bda3a-d05d-43dd-a265-dca0ad4f3f7a)

**Reply by joydeep-bhowmik** [reply]
Try v3.3.3 .... See if it fix your problem or you have to wait for the next release.

**Comment by danie-ramdhani** [comment]
fixed https://github.com/alpinejs/alpine/pull/4015 in [v3.4.5](https://github.com/livewire/livewire/releases/tag/v3.4.5)

**Reply by cbaconnier** [reply]
Thanks !
But note there's still some issues with components that could have modified the DOM (Flapickr is a good example) and not re-initialized it before returning back.

**Reply by ZermattChris** [reply]
I've just run into this issue as well (also Flatpickr).

These edge cases are killer :-/

Comforting to see others stuck on the same spot...

**Reply by cbaconnier** [reply]
@ZermattChris Concerning flatpickr you can use this trick https://github.com/livewire/livewire/discussions/6374#discussioncomment-7870909 But each cases will have it's own quirks. As for now, I decided to not use `wire:navigate` until I'm sure it won't be an issue as our application move forward.

**Reply by ZermattChris** [reply]
Have tried all the above (removed wire:navigate, wire:link & added instance.destroy(); as per the comment)...

Still blows up after 1-2 changing pages, with a 

```
Alpine Expression Error: instance is not defined

Expression: "instance.destroy();"

<div class="relative " x-data="datepicker(window.Livewi…angle('currDate').live)" x-on:livewire:navigating.window="instance.destroy();">
```

Too bad, thought I'd added a nice Date Range picker.
Thanks for the quick feedback -- I'm still finding my feet with Laravel, Livewire & Alpine. Would be utterly sunk without your inputs.

-Chris

**Reply by ZermattChris** [reply]
It looks like the following is also causing it to die:

```
$this->redirect("/new-booking/$this->currdate/$slotId/$timeHint", navigate: true);
```
Guess this is just more wire:navigate...

I've taken out all of these navigate:true bits and now things seem to be working! Yeay!!!

Huge thanks all -- I would *never* have figured this out on my own.

-Chris

**Comment by Jhamnerx** [comment]
I get a similar error, but it is using the wire/ui components
It only happens on Linux servers in Windows and it works perfectly.
livewire/livewire                   v3.4.9 
wireui/wireui                       v1.19.0 
![image](https://github.com/livewire/livewire/assets/30221273/a394a59e-3150-478d-a31a-3cb4b3816d00)

**Comment by briancalma** [comment]
Hey guys anyone here having this same issue?

**Reply by Fsmash** [reply]
https://alpinejs.dev/magics/store

Only work around I've found so far.

**Comment by Fsmash** [comment]
@joshhanley 

Sorry for pinging you but this has been an unresolved issue with no work around for quite some time now. Would utilzing global Alpine stores work as a workaround?

https://alpinejs.dev/magics/store
 
I will test on my own but for now I've just opted out of using wire:navigate altogether. Please let me know if there is some fix that I am just not aware of.

Edit: Just tested it out. It seems to fix things. I haven't really pushed the capabilities of doing it this way but I'll report back whatever issues I run into.

**Reply by Fsmash** [reply]
After trying it out some more, this seems like the fix. I would just use alpines global store if you can.

The weird part about this issue is that even if these expression errors get thrown everything still works for the most part. Navigation with the forwards and back button is what gets broken mainly with pages behaving kind of jank and not hydrating properly with new content after the navigate with the forward or back button. 

Is there any way for me to force the forwards and back page navigation to always just reload the page?

**Reply by joshhanley** [reply]
@Fsmash and anyone else in this thread, we're currently in the process of investigating all `wire:navigate` issues. I don't exactly follow what the issue is here. Any chance you could provide some details around it with information/ code about what is happening exactly (with screenshots/ screen capture if possible) and how you are expecting it to work instead. That'd be a great help!
